projects/cm_mcu/ZynqMonTask.c

```diff
@@ -268,17 +270,123 @@ void zm_set_firefly_temps(struct zynqmon_data_t data[], int start)
 {
   // Fireflies
   // update the data for ZMON
-  for (int i = 0; i < NFIREFLIES; i++) {
+  for (uint8_t i = 0; i < NFIREFLIES; i++) {
     data[i].sensor = i + start; // sensor id
     if (!isFFStale()) {
       data[i].data.i = getFFtemp(i); // sensor value and type
     }
     else {
       data[i].data.i = -56; // special stale value
     }
+    log_debug(LOG_SERVICE, " ff %d opt pow : 0x%02x\r\n", i, data[i].data.us);
   }
 }
 
+// updated once per loop. Store the firefly temperature and present-bit data
+void zm_set_firefly_bits(struct zynqmon_data_t data[], int start)
+{
+
+  int ll = 0;
+
+  for (int j = 0; j < NFIREFLY_ARG; ++j) { // loop over ff structs
+
+    uint16_t val;
+    if (isFFStale()) {
+      data[ll].data.us = 0xFF; // special stale value
+    }
+    else {
+      if (j == 0)
+        val = ((ffl12_f1_args.ffpart_bit_mask & 0x1) << 1) | (ffl12_f1_args.ffpart_bit_mask & 0x1) | ((ffl12_f1_args.ffpart_bit_mask & 0x2) << 2) | ((ffl12_f1_args.ffpart_bit_mask & 0x2) << 1) | ((ffl12_f1_args.ffpart_bit_mask & 0x4) << 3) | ((ffl12_f1_args.ffpart_bit_mask & 0x4) << 2);
+      else if (j == 2)
+        val = ((ffl12_f2_args.ffpart_bit_mask & 0x1) << 1) | (ffl12_f2_args.ffpart_bit_mask & 0x1) | ((ffl12_f2_args.ffpart_bit_mask & 0x2) << 2) | ((ffl12_f2_args.ffpart_bit_mask & 0x2) << 1) | ((ffl12_f2_args.ffpart_bit_mask & 0x4) << 3) | ((ffl12_f2_args.ffpart_bit_mask & 0x4) << 2);
+      else
```

Comment:
this code might need a comment or two to explain what it is doing. also a lot of magic numbers?

Reply:
Done.